applications/serial_lte_modem/src/slm_at_tcp_proxy.c

```diff
@@ -435,76 +452,199 @@ static int tcp_data_save(uint8_t *data, uint32_t length)
+		ret = accept(proxy.sock,
+				(struct sockaddr *)&remote, &len);
 		if (ret < 0) {
 			LOG_ERR("accept() failed: %d", -errno);
-			do_tcp_server_stop(-errno);
-			return;
+			return -errno;
 		}
+		LOG_DBG("accept(): %d", ret);
 		if (inet_ntop(AF_INET, &remote.sin_addr, peer_addr,
 			INET_ADDRSTRLEN) != NULL) {
 			sprintf(rsp_buf, "#XTCPSVR: %s connected\r\n",
 				peer_addr);
 			rsp_send(rsp_buf, strlen(rsp_buf));
 		}
 		proxy.sock_peer = ret;
+		LOG_DBG("New connection - %d",
+			proxy.sock_peer);
+		fds[nfds].fd = proxy.sock_peer;
+		fds[nfds].events = POLLIN;
+		nfds++;
 		/* Start a one-shot timer to close the connection */
-		k_timer_start(&conn_timer, K_SECONDS(CONFIG_SLM_TCP_CONN_TIME),
+		k_timer_start(&conn_timer,
+			      K_SECONDS(CONFIG_SLM_TCP_CONN_TIME),
 			      K_NO_WAIT);
-	}
-
-	if (proxy.role == AT_TCP_ROLE_SERVER) {
-		sock = proxy.sock_peer;
 	} else {
-		sock = proxy.sock;
+		char data[NET_IPV4_MTU];
+
+		ret = recv(fds[infd].fd, data, NET_IPV4_MTU, 0);
+		if (ret < 0) {
+			LOG_WRN("recv() error: %d", -errno);
+			return -errno;
+		}
+		if (proxy.datamode) {
+			rsp_send(data, ret);
+		} else if (slm_util_hex_check(data, ret)) {
+			ret = slm_util_htoa(data, ret, data_hex,
+				DATA_HEX_MAX_SIZE);
+			if (ret < 0) {
+				LOG_ERR("hex convert error: %d", ret);
+				return ret;
+			}
+			if (tcp_data_save(data_hex, ret) < 0) {
+				sprintf(rsp_buf,
+					"#XTCPDATA: overrun\r\n");
+			} else {
+				sprintf(rsp_buf,
+					"#XTCPDATA: %d, %d\r\n",
+					DATATYPE_HEXADECIMAL,
+					ret);
+			}
+			rsp_send(rsp_buf, strlen(rsp_buf));
+		} else {
+			if (tcp_data_save(data, ret) < 0) {
+				sprintf(rsp_buf,
+					"#XTCPDATA: overrun\r\n");
+			} else {
+				sprintf(rsp_buf,
+					"#XTCPDATA: %d, %d\r\n",
+					DATATYPE_PLAINTEXT, ret);
+			}
+			rsp_send(rsp_buf, strlen(rsp_buf));
+		}
+		/* Restart conn timer */
+		k_timer_stop(&conn_timer);
+		LOG_ERR("restart timer: POLLIN");
+		k_timer_start(&conn_timer,
+			      K_SECONDS(CONFIG_SLM_TCP_CONN_TIME),
+			      K_NO_WAIT);
 	}
-	fds.fd = sock;
-	fds.events = POLLIN;
+
+	return ret;
+}
+
+/* TCP server thread */
+static void tcpsvr_thread_func(void *p1, void *p2, void *p3)
+{
+	int ret, current_size;
+
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	fds[nfds].fd = proxy.sock;
+	fds[nfds].events = POLLIN;
+	nfds++;
 	ring_buf_reset(&data_buf);
 	while (true) {
 		if (proxy.role == AT_TCP_ROLE_SERVER &&
 			k_timer_status_get(&conn_timer) > 0) {
 			k_timer_stop(&conn_timer);
 			LOG_INF("Connecion timeout");
-			sprintf(rsp_buf, "#XTCPSVR: timeout\r\n");
+			sprintf(rsp_buf, "#XTCPSVR: %d timeout\r\n", -ETIME);
 			rsp_send(rsp_buf, strlen(rsp_buf));
 			close(proxy.sock_peer);
-			goto thread_entry;
+			proxy.sock_peer = INVALID_SOCKET;
 		}
-		ret = poll(&fds, 1, MSEC_PER_SEC * CONFIG_SLM_TCP_POLL_TIME);
+		ret = poll(fds, nfds, MSEC_PER_SEC * CONFIG_SLM_TCP_POLL_TIME);
+		if (ret < 0) {  /* IO error */
+			LOG_WRN("poll() error: %d", ret);
+			return;
```

Comment:
Who does the cleanup in such case? We might end up with two dangling sockets.
I think we should loop over `fds` and close any valid socket in case of errors (i. e. if we finalize the thread execution). They're no longer be used anyway.

Reply:
All the clean up tasks are now handled in the exit: of thread.